gaia-header.js

```diff
@@ -34,124 +54,519 @@ module.exports = Component.register('gaia-header', {
    */
   created: function() {
-    this.createShadowRoot().innerHTML = this.template;
+    debug('created');
+    this.setupShadowRoot();
 
-    // Get els
+    // Elements
     this.els = {
       actionButton: this.shadowRoot.querySelector('.action-button'),
-      headings: this.querySelectorAll('h1,h2,h3,h4'),
-      inner: this.shadowRoot.querySelector('.inner')
+      buttons: this.querySelectorAll('button, a'),
+      titles: this.querySelectorAll('h1')
     };
 
+    // Events
     this.els.actionButton.addEventListener('click', e => this.onActionButtonClick(e));
-    this.configureActionButton();
-    this.runFontFit();
+    this.observer = new MutationObserver(this.onMutation.bind(this));
+
+    // Properties
+    this.titleEnd = this.getAttribute('title-end');
+    this.titleStart = this.getAttribute('title-start');
+    this.noFontFit = this.getAttribute('no-font-fit');
+    this.notFlush = this.hasAttribute('not-flush');
+    this.action = this.getAttribute('action');
+
+    this.unresolved = {};
+    this.pending = {};
   },
 
   /**
    * Called when the element is
    * attached to the DOM.
    *
+   * Run fontFit when we have DOM
+   * context and start listening
+   * for DOM mutations.
+   *
+   * We run font-fit on next tick to
+   * avoid reading from the DOM when
+   * it may not be loaded yet.
+   *
    * @private
    */
   attached: function() {
-    this.rerunFontFit();
+    debug('attached');
+    this.runFontFitSoon();
+    this.observerStart();
   },
 
   /**
-   * Called when one of the attributes
-   * on the element changes.
+   * Called when the element is
+   * detached from the DOM.
    *
    * @private
    */
-  attributeChanged: function(attr) {
-    if (attr === 'action') {
-      this.configureActionButton();
-      this.rerunFontFit();
-    }
+  detached: function() {
+    debug('detached');
+    this.observerStop();
+    this.clearPending();
   },
 
   /**
-   * Runs the logic to size and position
-   * header text inside the available space.
+   * Clears pending `.nextTick()`s
    *
    * @private
    */
-  runFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
-      fontFit.observeHeadingChanges(this.els.headings[i]);
+  clearPending: function() {
+    for (var key in this.pending) {
+      this.pending[key].clear();
+      delete this.pending[key];
     }
   },
 
   /**
-   * Rerun font-fit logic.
+   * Run the font-fit logic and
+   * center the title.
+   *
+   * The styles are calculated synchronously,
+   * but then set asynchronously. This means
+   * this function can be hammered in one turn,
+   * and the title styles will only be written
+   * once, with the latest styles.
+   *
+   * @return {Promise}
+   * @public
+   */
+  runFontFit: function() {
+    debug('run font-fit');
+
+    // Nothing is run if `no-font-fit` attr is present
+    if (this.noFontFit) { return Promise.reject('no-font-fit'); }
```

Comment:
Didn't we agree earlier that this should `resolve()` as it's not an 'error' state.